tss-esapi/src/context/tpm_commands/session_commands.rs

```diff
@@ -59,6 +59,8 @@ impl Context {
         auth_hash: HashingAlgorithm,
     ) -> Result<Option<AuthSession>> {
         let mut session_handle = ObjectHandle::None.into();
+        let some_nonce = nonce.is_some();
+        let tss_nonce = nonce.map(TPM2B_NONCE::from).unwrap_or_default();
```

Comment:
You can also do like this

```rust
        let potential_tpm2b_nonce = nonce.map(|v| v.into());
        let tpm2b_nonce_ptr = potential_tpm2b_nonce.as_ref().map_or_else(null, |v| v);
```

If you want to avoid importing the TPM2B_NONCE type.

Reply:
it would also be possible of course to combine them as long as it is outside the unsafe block then rust can determine in what scope the value must be in order to be valid.

```rust
let tpm2b_nonce_ptr = nonce.map(|v| v.into()).as_ref().map_or_else(null, |v| v);
```

Reply:
@Superhepper do you have a way to reproducibly test if the leak is fixed or not (either a sample project with valgrind or something like that). I'd like to experiment with different ways to express this PR? (I'm OK with merging this as is if you say it's good, just curious about further improvements).

Reply:
Yes. I just installed cargo-valgrind and ran all tests with that before and after this fix. And before the fix you get an endless amount of complaints about how conditional jumps depends on uninitialized values.

```
<error>
  <unique>0x30</unique>
  <tid>1</tid>
  <kind>UninitCondition</kind>
  <what>Conditional jump or move depends on uninitialised value(s)</what>
  <stack>
    <frame>
      <ip>0x484027E</ip>
      <obj>/usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so</obj>
      <fn>memcpy@GLIBC_2.2.5</fn>
    </frame>
    <frame>
      <ip>0x4911DD6</ip>
      <obj>/usr/local/lib/libtss2-mu.so.0.0.0</obj>
      <fn>Tss2_MU_TPM2B_NONCE_Marshal</fn>
    </frame>
    <frame>
      <ip>0x4FB9A7D</ip>
      <obj>/usr/local/lib/libtss2-sys.so.1.0.0</obj>
      <fn>Tss2_Sys_StartAuthSession_Prepare</fn>
    </frame>
    <frame>
      <ip>0x48B4C0F</ip>
      <obj>/usr/local/lib/libtss2-esys.so.0.0.0</obj>
      <fn>Esys_StartAuthSession_Async</fn>
    </frame>
    <frame>
      <ip>0x48B465F</ip>
      <obj>/usr/local/lib/libtss2-esys.so.0.0.0</obj>
      <fn>Esys_StartAuthSession</fn>
    </frame>
    <frame>
      <ip>0x646042</ip>
      <obj>/home/hepper/workspace/rust-tss-esapi/target/debug/deps/integration_tests-4b1afde679bf38f9</obj>
      <fn>tss_esapi::context::tpm_commands::session_commands::&lt;impl tss_esapi::context::Context&gt;::start_auth_session</fn>
      <dir>/home/hepper/workspace/rust-tss-esapi/tss-esapi/src/context/tpm_commands</dir>
      <file>session_commands.rs</file>
      <line>64</line>
    </frame>
    <frame>
      <ip>0x3AECE8</ip>
      <obj>/home/hepper/workspace/rust-tss-esapi/target/debug/deps/integration_tests-4b1afde679bf38f9</obj>
      <fn>integration_tests::context_tests::tpm_commands::session_commands_tests::test_start_auth_session::test_nonce_sess</fn>
      <dir>/home/hepper/workspace/rust-tss-esapi/tss-esapi/tests/integration_tests/context_tests/tpm_commands</dir>
      <file>session_commands_tests.rs</file>
      <line>31</line>
    </frame>
    <frame>
      <ip>0x444669</ip>
      <obj>/home/hepper/workspace/rust-tss-esapi/target/debug/deps/integration_tests-4b1afde679bf38f9</obj>
      <fn>integration_tests::context_tests::tpm_commands::session_commands_tests::test_start_auth_session::test_nonce_sess::{{closure}}</fn>
      <dir>/home/hepper/workspace/rust-tss-esapi/tss-esapi/tests/integration_tests/context_tests/tpm_commands</dir>
      <file>session_commands_tests.rs</file>
      <line>29</line>
    </frame>
    <frame>
      <ip>0x4950FD</ip>
      <obj>/home/hepper/workspace/rust-tss-esapi/target/debug/deps/integration_tests-4b1afde679bf38f9</obj>
      <fn>core::ops::function::FnOnce::call_once</fn>
      <dir>/rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/core/src/ops</dir>
      <file>function.rs</file>
      <line>227</line>
    </frame>
    <frame>
      <ip>0x58DB62</ip>
      <obj>/home/hepper/workspace/rust-tss-esapi/target/debug/deps/integration_tests-4b1afde679bf38f9</obj>
      <fn>call_once&lt;fn(), ()&gt;</fn>
      <dir>/rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/core/src/ops</dir>
      <file>function.rs</file>
      <line>227</line>
    </frame>
    <frame>
      <ip>0x58DB62</ip>
      <obj>/home/hepper/workspace/rust-tss-esapi/target/debug/deps/integration_tests-4b1afde679bf38f9</obj>
      <fn>test::__rust_begin_short_backtrace</fn>
      <dir>library/test/src</dir>
      <file>lib.rs</file>
      <line>585</line>
    </frame>
    <frame>
      <ip>0x58C5E3</ip>
      <obj>/home/hepper/workspace/rust-tss-esapi/target/debug/deps/integration_tests-4b1afde679bf38f9</obj>
      <fn>call_once&lt;(), (dyn core::ops::function::FnOnce&lt;(), Output=()&gt; + core::marker::Send), alloc::alloc::Global&gt;</fn>
      <dir>/rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/alloc/src</dir>
      <file>boxed.rs</file>
      <line>1694</line>
    </frame>
    <frame>
      <ip>0x58C5E3</ip>
      <obj>/home/hepper/workspace/rust-tss-esapi/target/debug/deps/integration_tests-4b1afde679bf38f9</obj>
      <fn>call_once&lt;(), alloc::boxed::Box&lt;(dyn core::ops::function::FnOnce&lt;(), Output=()&gt; + core::marker::Send), alloc::alloc::Global&gt;&gt;</fn>
      <dir>/rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/core/src/panic</dir>
      <file>unwind_safe.rs</file>
      <line>271</line>
    </frame>
    <frame>
      <ip>0x58C5E3</ip>
      <obj>/home/hepper/workspace/rust-tss-esapi/target/debug/deps/integration_tests-4b1afde679bf38f9</obj>
      <fn>do_call&lt;core::panic::unwind_safe::AssertUnwindSafe&lt;alloc::boxed::Box&lt;(dyn core::ops::function::FnOnce&lt;(), Output=()&gt; + core::marker::Send), alloc::alloc::Global&gt;&gt;, ()&gt;</fn>
      <dir>/rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src</dir>
      <file>panicking.rs</file>
      <line>406</line>
    </frame>
    <frame>
      <ip>0x58C5E3</ip>
      <obj>/home/hepper/workspace/rust-tss-esapi/target/debug/deps/integration_tests-4b1afde679bf38f9</obj>
      <fn>try&lt;(), core::panic::unwind_safe::AssertUnwindSafe&lt;alloc::boxed::Box&lt;(dyn core::ops::function::FnOnce&lt;(), Output=()&gt; + core::marker::Send), alloc::alloc::Global&gt;&gt;&gt;</fn>
      <dir>/rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src</dir>
      <file>panicking.rs</file>
      <line>370</line>
    </frame>
    <frame>
      <ip>0x58C5E3</ip>
      <obj>/home/hepper/workspace/rust-tss-esapi/target/debug/deps/integration_tests-4b1afde679bf38f9</obj>
      <fn>catch_unwind&lt;core::panic::unwind_safe::AssertUnwindSafe&lt;alloc::boxed::Box&lt;(dyn core::ops::function::FnOnce&lt;(), Output=()&gt; + core::marker::Send), alloc::alloc::Global&gt;&gt;, ()&gt;</fn>
      <dir>/rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src</dir>
      <file>panic.rs</file>
      <line>133</line>
    </frame>
    <frame>
      <ip>0x58C5E3</ip>
      <obj>/home/hepper/workspace/rust-tss-esapi/target/debug/deps/integration_tests-4b1afde679bf38f9</obj>
      <fn>run_test_in_process</fn>
      <dir>library/test/src</dir>
      <file>lib.rs</file>
      <line>608</line>
    </frame>
    <frame>
      <ip>0x58C5E3</ip>
      <obj>/home/hepper/workspace/rust-tss-esapi/target/debug/deps/integration_tests-4b1afde679bf38f9</obj>
      <fn>test::run_test::run_test_inner::{{closure}}</fn>
      <dir>library/test/src</dir>
      <file>lib.rs</file>
      <line>500</line>
    </frame>
    <frame>
      <ip>0x58B964</ip>
      <obj>/home/hepper/workspace/rust-tss-esapi/target/debug/deps/integration_tests-4b1afde679bf38f9</obj>
      <fn>test::run_test::run_test_inner</fn>
      <dir>/rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b//library/test/src</dir>
      <file>lib.rs</file>
      <line>538</line>
    </frame>
  </stack>
</error>
```

Reply:
Great, thanks! :+1:

Reply:
It would be really good if we could have this check in our CI as discussed here 
https://github.com/parallaxsecond/rust-tss-esapi/issues/339

Reply:
Hmm, about your last suggestion - I'm a bit worried given the [reference documentation](https://doc.rust-lang.org/stable/reference/destructors.html#temporary-lifetime-extension) saying:

> If a borrow, dereference, field, or tuple indexing expression has an extended temporary scope then so does its operand. If an indexing expression has an extended temporary scope then the indexed expression also has an extended temporary scope.

My worry is that holding on to a pointer derived from a reference might not be enough. I'm tempted to go somewhere in-between, storing the `potential_tpm2b_nonce` and then converting it to a pointer in the call block. Does that make sense?

Reply:
Yeah, it does. Because then you will, without doubt, bind the converted value to the method scope instead of having to rely on rusts rather confusing rules regarding extended life times for temporary objects.

Reply:
Done